pippy/PipelineSchedule.py

```diff
@@ -42,6 +43,21 @@ def __init__(
         )
         self._output_merge_spec = output_merge_spec
 
+    def _configure_data_parallel_mode(self, last_backward: bool):
+        """
+        Whether using PP with FSDP or DDP, there are some runtime differences between the last backward step and the
+        other steps.  Namely, we need to accumulate gradients on previous steps and reduce them on the last step, but
+        there are additional state-variables and performance considerations depending on the data parallelism used.
+
+        This helper should adapt any pipeline parallel schedule to work with common/supported data parallel libraries.
+        """
+        module = self._stage.submod
+        if isinstance(module, FSDP):
+            module.set_is_last_backward(last_backward)
+            module.set_requires_gradient_sync(last_backward)
+        else:
+            raise NotImplementedError("Only FSDP is supported now, please add support for your data parallel library")
+
```

Comment:
This looks simple and clean.
nit: do you mind moving this method to `PipelineStage` level? 
Two reasons:
1. we need to access `submod` here;
2. I am refactoring self._stage -- in case of multiple stages per rank, `self._stage` will be gone (and you probably don't want two implementations here).

Reply:
In `PipelineStage`, you can get `last_backward: bool` by checking `bwd_chunk_id`, like here:
https://github.com/pytorch/PiPPy/blob/9cd29d17324a53b1342c2b6fc5eeef5f5405ec49/pippy/PipelineStage.py#L149-L152

Reply:
yep, that sounds good!

Reply:
@H-Huang can you TAL? 

I just realized that there is already DDP support in there in PipelineStageBase in the form of the _nosync methods.

Do you think it would be easy to put the ddp nosync into the 'configure_data_parallel_mode` helper and then remove the 'maybe_with_nosync' helpers, or should we put FSDP into those helpers instead?

Reply:
I think we should put the ddp nosync logic into the `'configure_data_parallel_mode`  method you created. I think it is clearer and more explicit that the forward of the module is being modified. I'm happy to do this in a follow up issue or PR as well!

Reply:
ok. in this PR i just moved the `_configure_data_parallel_mode` helper into stage for now.

we may need to do several changes to migrate DDP, so lets keep that in a separate PR and also add DDP to the unit test for that one.  I'll let you do that if you are offering.